src/common/MemoryModel.cc

```diff
@@ -72,35 +99,47 @@ std::optional<int64_t> MemoryModel::get_mapped_heap()
 }
 
 
-
-void MemoryModel::_sample(mem_snap_t *psnap)
+std::optional<mem_snap_t> MemoryModel::full_sample()
 {
   if (!proc_status.is_open()) {
-    ldout(cct, 0) <<  fmt::format("MemoryModel::sample() unable to open {}", proc_stat_fn) << dendl;
-    return;
+    ldout(cct, 0) << fmt::format(
+			 "MemoryModel::sample() unable to open {}",
+			 proc_stat_fn)
+		  << dendl;
+    return std::nullopt;
   }
   // always rewind before reading
   proc_status.clear();
   proc_status.seekg(0);
 
-  while (!proc_status.eof()) {
-    string line;
-    getline(proc_status, line);
-
-    if (strncmp(line.c_str(), "VmSize:", 7) == 0)
-      psnap->size = atol(line.c_str() + 7);
-    else if (strncmp(line.c_str(), "VmRSS:", 6) == 0)
-      psnap->rss = atol(line.c_str() + 7);
-    else if (strncmp(line.c_str(), "VmHWM:", 6) == 0)
-      psnap->hwm = atol(line.c_str() + 7);
-    else if (strncmp(line.c_str(), "VmLib:", 6) == 0)
-      psnap->lib = atol(line.c_str() + 7);
-    else if (strncmp(line.c_str(), "VmPeak:", 7) == 0)
-      psnap->peak = atol(line.c_str() + 7);
-    else if (strncmp(line.c_str(), "VmData:", 7) == 0)
-      psnap->data = atol(line.c_str() + 7);
+  mem_snap_t s;
+  // we will be looking for 6 entries
+  int yet_to_find = 6;
+
+  while (!proc_status.eof() && yet_to_find > 0) {
+    string ln;
+    getline(proc_status, ln);
+
+    if (cmp_against(ln, "VmSize:", s.size) ||
+	cmp_against(ln, "VmRSS:", s.rss) || cmp_against(ln, "VmHWM:", s.hwm) ||
+	cmp_against(ln, "VmLib:", s.lib) ||
+	cmp_against(ln, "VmPeak:", s.peak) ||
+	cmp_against(ln, "VmData:", s.data)) {
```

Comment:
Would it be too much to push all required items in a hash table and test for the presence of the prefix of the current line in the hash table instead ?

Reply:
@mchangir ,
> Would it be too much to push all required items in a hash table and test for the presence of the prefix of the current line in the hash table instead ?

I do not see a way to do have better performance this way (I had originally measured multiple ideas I could think of). For example - note that we have sub-strings of at least two lengths here.